### PR TITLE
SfM from .obj files

### PR DESCRIPTION
**Pull request by prajwalchidananda**
New features:
- Create `look-at-poses`
- Render images from `.obj` files using pytorch3d
- Create colmap sfm from synthetic poses and images

### CODE REVIEW

File: pixtrack/utils/pytorch3d_render_utils.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                             R=R, 
+                             T=T,
+                             in_ndc=False,
+                             image_size=(image_size,))
+
+    # Rasterization Settings
+    raster_settings = RasterizationSettings(
+                             image_size=image_size, 
+                             blur_radius=0.0, 
+                             faces_per_pixel=1,)
```
Comment (saunair) [inline pixtrack/utils/pytorch3d_render_utils.py]: Doesn't this depend on the mesh resolution? Given we're working with high res, shouldn't we track more? (as the faces are pretty small)

File: pixtrack/utils/pytorch3d_render_utils.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                                     radius=radius, 
+                                     color=None)
+    v = sph.vertices
+    Rs = [look_at_rotation(torch.tensor(v[i]).unsqueeze(0).float()).squeeze().numpy() for i in range(v.shape[0])]
+    Rs = [R.from_matrix(x).as_matrix() for x in Rs]
+    Ts = [-Rs[i].T @ v[i] for i in range(v.shape[0])]
+    Rs = torch.tensor(np.stack(Rs))
+    Ts = torch.tensor(np.stack(Ts))
+    return Rs, Ts
+
```
Comment (saunair) [inline pixtrack/utils/pytorch3d_render_utils.py:33]: Nit: We should start adding docstrings as we mature this code base. Makes the usability a lot higher post that

File: pixtrack/utils/pytorch3d_render_utils.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+def create_look_at_poses_for_mesh(mesh_path, subdivisions=2):
+    if torch.cuda.is_available():
+        device = torch.device("cuda:0")
```
Comment (saunair) [inline pixtrack/utils/pytorch3d_render_utils.py]: make this a param? We can expose it later

File: pixtrack/utils/pytorch3d_render_utils.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                             image_size=image_size, 
+                             blur_radius=0.0, 
+                             faces_per_pixel=1,)
+
+    # Create lights
+    lights = AmbientLights(device=device)
+
+    # Create renderer
+    renderer = MeshRenderer(
+                            rasterizer=MeshRasterizer(
```
Comment (saunair) [inline pixtrack/utils/pytorch3d_render_utils.py]: I have some style comments, but I'll be that guy once we have finalized the paper :)

File: pixtrack/utils/pytorch3d_render_utils.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    # Render image
+    image = renderer(mesh).squeeze().cpu().numpy()
+
+    return image
+
+def create_colmap_image_from_pytorch3d_RT(R, T, image_name, image_id, camera_id):
+    
+    # Flip x, y axes
+    flip = np.eye(3)
+    flip[0, 0] = -1.
```
Comment (saunair) [inline pixtrack/utils/pytorch3d_render_utils.py]: 2 lines of code worth your weekend

File: scripts/create_sfm_from_obj.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    import_features(image_ids, database, features)
+    import_matches(image_ids, database, pairs, matches,
+                   min_match_score=None, 
+                   skip_geometric_verification=False)
+    geometric_verification(database, pairs, verbose=True)
+
+    # Create images, cameras and points3D binaries
+    write_images_binary(colmap_images, raw_dir / 'images.bin')
+    write_cameras_binary(colmap_cameras, raw_dir / 'cameras.bin')
+    write_points3D_binary({}, raw_dir / 'points3D.bin') #Should be empty
```
Comment (saunair) [inline scripts/create_sfm_from_obj.py]: How hard would it be to support this? Given I might need 3d points for the depth stuff. 
triangulation add that info to the dir?

File: scripts/create_sfm_from_obj.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                  image_list=None)
+    image_ids = get_image_ids(database)
+    import_features(image_ids, database, features)
+    import_matches(image_ids, database, pairs, matches,
+                   min_match_score=None, 
+                   skip_geometric_verification=False)
+    geometric_verification(database, pairs, verbose=True)
+
+    # Create images, cameras and points3D binaries
+    write_images_binary(colmap_images, raw_dir / 'images.bin')
```
Comment (saunair) [inline scripts/create_sfm_from_obj.py]: good way to circumvent BA

File: scripts/create_sfm_from_obj.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    cx = 512. * 3
+    cy = 512. * 3
+    W = 1024 * 3
+    H = 1024 * 3
+    dataset_path = args.dataset_path
+    output_path = args.output_path
+    mesh_path = args.mesh_path
+
+    # Render images from obj, get colmap images and cameras
+    render_dataset_and_get_colmap_images_and_cameras(mesh_path, dataset_path,
```
Comment (saunair) [inline scripts/create_sfm_from_obj.py]: return not captured to pass to the next func

File: scripts/create_sfm_from_obj.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    matcher_conf['model']['weights'] = 'indoor'
+    references = [str(p.relative_to(images)) for p in (images / 'mapping/').iterdir()]
+    print(len(references), "mapping images")
+    
+    # Extract features and matches
+    extract_features.main(feature_conf, images, image_list=references, feature_path=features)
+    pairs_from_exhaustive.main(sfm_pairs, image_list=references)
+    match_features.main(matcher_conf, sfm_pairs, features=features, matches=matches)
+    pairs = sfm_pairs
+    assert features.exists(), features
```
Comment (saunair) [inline scripts/create_sfm_from_obj.py]: Generally nice to add a message with the assert. eg: ```f"Keypoint Feature extraction failed. Got {features} as features."```

File: scripts/create_sfm_from_obj.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    assert pairs.exists(), pairs
+    assert matches.exists(), matches
+
+    # Create a database and import images, features and matches
+    raw_dir.mkdir(parents=True, exist_ok=True)
+    database = raw_dir / 'database.db'
+    create_empty_db(database)
+    camera_mode = 'AUTO'
```
Comment (saunair) [inline scripts/create_sfm_from_obj.py]: Pull out as an arg to the func. Also what does this mean?

File: scripts/create_sfm_from_obj.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+def render_dataset_and_get_colmap_images_and_cameras(mesh_path, dataset_path, 
+                                                     fx, fy, cx, cy, W, H):
+    mesh_path = args.mesh_path
+    Rs, Ts, mesh = create_look_at_poses_for_mesh(mesh_path, subdivisions=2)
+    dataset_path.mkdir(parents=True, exist_ok=True)
+    mapping_path = dataset_path / 'mapping'
+    mapping_path.mkdir(parents=True, exist_ok=True)
+    colmap_images = {}
+    colmap_camera_id = 1
+    colmap_camera = create_colmap_camera(W, H, fx, cx, cy, k1=0.)
```
Comment (saunair) [inline scripts/create_sfm_from_obj.py]: pull the K out as a default arg to the func

File: scripts/create_sfm_from_obj.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        T = Ts[i].unsqueeze(0)
+        image = render_image(mesh, fx, fy, cx, cy, W, H, R, T)
+        image = (image[..., :3] * 255.).astype(np.uint8)
+        image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)
+        img_name = ('%d.png' % (i + 1)).zfill(8)
+        cv2.imwrite(str(mapping_path / img_name), image)
+
+        colmap_image_name = 'mapping/' + img_name
+        colmap_image_id = i + 1
+        colmap_image = create_colmap_image_from_pytorch3d_RT(R[0], T[0],
```
Comment (saunair) [inline scripts/create_sfm_from_obj.py]: Indexing for `0` is hard to interpret. Can you make it functional instead?

File: scripts/create_sfm_from_obj.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                                 get_image_ids)
+from hloc.utils.read_write_model import (write_images_binary, 
+                                         write_points3D_binary,
+                                         write_cameras_binary)
+
+
+def render_dataset_and_get_colmap_images_and_cameras(mesh_path, dataset_path, 
+                                                     fx, fy, cx, cy, W, H):
+    mesh_path = args.mesh_path
+    Rs, Ts, mesh = create_look_at_poses_for_mesh(mesh_path, subdivisions=2)
```
Comment (saunair) [inline scripts/create_sfm_from_obj.py]: subdivisions needs to be a default arg of the func

**Review by saunair — COMMENTED** [review]
Have a few comments. I think a bug got added while cleaning up in `create_sfm_from_obj` 
This tool itself is useful outside of pixtrack. I think eventually we can make your SFM code itself as an independent repo

File: pixtrack/utils/pytorch3d_render_utils.py (hunk 2 above)
Comment (prajwalchidananda) [reply pixtrack/utils/pytorch3d_render_utils.py:33]: Agreed

File: pixtrack/utils/pytorch3d_render_utils.py (hunk 1 above)
Comment (prajwalchidananda) [reply pixtrack/utils/pytorch3d_render_utils.py]: The renderer got pretty slow when I tried increasing this value. Plus, the outputs seem pretty sharp and high res even with this setting

File: pixtrack/utils/pytorch3d_render_utils.py (hunk 3 above)
Comment (prajwalchidananda) [reply pixtrack/utils/pytorch3d_render_utils.py]: Done

File: pixtrack/utils/pytorch3d_render_utils.py (hunk 4 above)
Comment (prajwalchidananda) [reply pixtrack/utils/pytorch3d_render_utils.py]: I've started using flake8 and black with 88 as the default line width

File: pixtrack/utils/pytorch3d_render_utils.py (hunk 5 above)
Comment (prajwalchidananda) [reply pixtrack/utils/pytorch3d_render_utils.py]: Sigh

File: scripts/create_sfm_from_obj.py (hunk 9 above)
Comment (prajwalchidananda) [reply scripts/create_sfm_from_obj.py]: Done

File: scripts/create_sfm_from_obj.py (hunk 11 above)
Comment (prajwalchidananda) [reply scripts/create_sfm_from_obj.py]: Done

File: scripts/create_sfm_from_obj.py (hunk 12 above)
Comment (prajwalchidananda) [reply scripts/create_sfm_from_obj.py]: Refactored the code

File: scripts/create_sfm_from_obj.py (hunk 13 above)
Comment (prajwalchidananda) [reply scripts/create_sfm_from_obj.py]: Done

File: scripts/create_sfm_from_obj.py (hunk 6 above)
Comment (prajwalchidananda) [reply scripts/create_sfm_from_obj.py]: Triangulation (the next step) fills `points3D.bin` with 3d points. The SfM created is identical to the real SfMs we've created as part of our initial experiments.

File: scripts/create_sfm_from_obj.py (hunk 8 above)
Comment (prajwalchidananda) [reply scripts/create_sfm_from_obj.py]: Fixed

File: scripts/create_sfm_from_obj.py (hunk 10 above)
Comment (prajwalchidananda) [reply scripts/create_sfm_from_obj.py]: These are camera modes in pycolmap. I changed this to using the enum name, rather than hardcoding the string.

**Comment by prajwalchidananda** [comment]
@saunair Resolved all comments and tested the code before pushing.

**Review by saunair — APPROVED** [review]
Thanks for addressing

**Pull request merged**